### DISCUSSION TITLE
What is the best tool or ORM to manage database in Fast API?

### DISCUSSION THREAD

**Discussion by Achilles0509** [fastapi/fastapi#8615] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import List

import databases
import sqlalchemy
from fastapi import FastAPI
from pydantic import BaseModel

# SQLAlchemy specific code, as with any other app
DATABASE_URL = "sqlite:///./test.db"
# DATABASE_URL = "postgresql://user:password@postgresserver/db"

database = databases.Database(DATABASE_URL)

metadata = sqlalchemy.MetaData()

notes = sqlalchemy.Table(
    "notes",
    metadata,
    sqlalchemy.Column("id", sqlalchemy.Integer, primary_key=True),
    sqlalchemy.Column("text", sqlalchemy.String),
    sqlalchemy.Column("completed", sqlalchemy.Boolean),
)


engine = sqlalchemy.create_engine(
    DATABASE_URL, connect_args={"check_same_thread": False}
)
metadata.create_all(engine)


class NoteIn(BaseModel):
    text: str
    completed: bool


class Note(BaseModel):
    id: int
    text: str
    completed: bool


app = FastAPI()


@app.on_event("startup")
async def startup():
    await database.connect()


@app.on_event("shutdown")
async def shutdown():
    await database.disconnect()


@app.get("/notes/", response_model=List[Note])
async def read_notes():
    query = notes.select()
    return await database.fetch_all(query)


@app.post("/notes/", response_model=Note)
async def create_note(note: NoteIn):
    query = notes.insert().values(text=note.text, completed=note.completed)
    last_record_id = await database.execute(query)
    return {**note.dict(), "id": last_record_id}
```

### Description

I read FastAPI documentation, but still not clear which tools should use to manage database in FastAPI projects.
encode/databases + SqlAlchemy or tortoise-orm?

### Operating System

Linux, Windows

### FastAPI Version

0.75.0

### Python Version

3.7 +

**Comment by yinziyan1206** [comment]
I use SqlModel and SqlAlchemy Core. I think it better to use the official ORM tool for FastAPI :)

**Comment by Achilles0509** [comment]
Yes, I think so.
But I don't think FastAPI has official ORM.
It's developed as micro web framework not full like Django or Laravel.

**Comment by yinziyan1206** [comment]
eh tiangolo created ``SqlModel``. It combines pydantic model and sqlalchemy and it's comvenient. If you want performance just use tortoise-orm.

**Comment by Kludex** [comment]
Using the async `Core SQLAlchemy` performs better than `TortoiseORM`. You can run the benchmark yourself [here](https://github.com/tortoise/orm-benchmarks/pull/11).

**Reply by Priyansh2** [reply]
@Kludex  which is the best library for async database connection which supports connection pooling. @tiangolo  mention databases in the fastapi docs. Is it better or there is some other available ?

**Reply by Kludex** [reply]
I use SQLAlchemy.

**Reply by Priyansh2** [reply]
@Kludex I tried databases using aiomysql driver and sent over 50k concurrent API requests using `asyncio` and `aiohttp`. My API reads MySQL database and performs some operations of complexity O(1). At some point, it throws `connection refused to 127.0.0.1 error`. After that, it failed again when I ran the program and sent 20K requests. I believe there were some hanging connections. I thought `databases` had garbage collection, but it doesn't. Finally, deciding to go with the async session module of sqlalchemy as it didn't fail by sending the same number of concurrent requests repeatedly. Even there, I found out I can't configure the pool via pool_size, pool_recycle, etc, as it throws the error. Anyway, thanks for your time.

**Comment by Achilles0509** [comment]
OK, I will check more about it.

My concern is tortoise-orm is growing faster than SqlAlchemy.
I think ranking, star etc are very important.
FastAPI is great framework, so it is growing faster than Django or Flask.
There will be reason why tortoise-orm is growing faster.

**Comment by yinziyan1206** [comment]
> Using the async `Core SQLAlchemy` performs better than `TortoiseORM`. You can run the benchmark yourself [here](https://github.com/tortoise/orm-benchmarks/pull/11).

that's awesome

**Comment by Achilles0509** [comment]
I made benchmark test on my PC.
But async `SqlAlchemy Core` is slower than sync `SqlAlchemy Core` and of course slower than `Tortoise ORM`

You can see result [here](https://github.com/tortoise/orm-benchmarks/pull/11#issuecomment-1075543926)

**Comment by Leon0824** [comment]
Try [Prisma](https://prisma-client-py.readthedocs.io/en/stable/).

Define your schema in its Prisma language, and use its generator to generate whole pydantic / db hybrid models. Then import the models module into your project.

**Reply by Filsommer** [reply]
I've read the prisma-python documentation and it is still not clear how to programatically convert a prisma model (e.g.)
```
model Item {
  id String    @id @default(cuid()) @db.VarChar(255)
  name String
  price Int
  created_at DateTime?
}
```

to a pydantic model, like:
```
class Item(BaseModel):
    id: str
    name: str
    price: int
    created_at: Optional[datetime.datetime] = datetime.datetime.now()
```

How can this be achieved?

**Comment by anthonycepeda** [comment]
Try [SQLModel](https://sqlmodel.tiangolo.com/). It’s Pydantic + SQLAlchemy.

Here's an [example](https://github.com/anthonycepeda/fastapi-sqlmodel)

**Comment by Kludex** [comment, marked as answer]
@tiangolo would say to check [SQLModel](https://sqlmodel.tiangolo.com).

But for the sake of answering the question... There's no "best tool or ORM to manage database in FastAPI". FastAPI works well with any tool, since it's not coupled with any.

**Comment by lozzo** [comment]
DONT USE  tortoise-orm !!! IT'S ORM BUILT ON BUGS

**Reply by enchance** [reply]
> DONT USE tortoise-orm !!! IT'S ORM BUILT ON BUGS

I don't think it's that bad in fact I much prefer the syntax of Tortoise-ORM. But I stopped using it because its migrations tool [Aerich](https://github.com/tortoise/aerich) is absolute garbage. The bugs Aerich causes (consistently) offset the usability of Tortoise which sad. I see myself using it again once Aerich realizes it's not all right in the head.

**Comment by sinisaos** [comment]
[Piccolo ORM](https://github.com/piccolo-orm/piccolo) is an easy-to-use (both `async` and `sync`) ORM and query builder that works great with FastAPI. It has automatic migrations ([they work pretty well and for some egde cases it's easy to use raw sql in migrations](https://github.com/piccolo-orm/piccolo/discussions/1170#discussioncomment-12617130)), authentication, [admin interface](https://github.com/piccolo-orm/piccolo_admin), [useful middleware](https://github.com/piccolo-orm/piccolo_api) etc. Pydantic can be used in two ways: to write the models ourselves or we can use the [create_pydantic_model](https://piccolo-orm.readthedocs.io/en/latest/piccolo/serialization/index.html#create-pydantic-model) utility function to convert a `Piccolo` table into a `Pydantic` model. I've done [some benchmarks](https://github.com/sinisaos/async-orm-benchmarks) for various async ORMs, query builders and drivers. In these tests I focus on `read data` (I didn't test inserts, updates and deletes) which I think is the main bottleneck in using ORM. That benchmarks (inspired by [Tortoise ORM benchmarks](https://github.com/tortoise/orm-benchmarks)) use a FastAPI web app with various ORMs because I think ORMs are mostly used in web applications. I don't know if these benchmarks are good enough, but tests show that Piccolo has better performance than other ORMs in my case.

**Comment by fmelihh** [comment]
Alembic and async sqlalchemy could be a good choice. I haven't seen tools like Tortoise ORM support less commonly used databases very well. For example, in the company I'm currently working at, we use SQLAlchemy as the core and support it with additional extensions like clickhouse sqlalchemy.  When choosing a package, it will be more appropriate to also determine the database you're going to use.

**Comment by vsdudakov** [comment]
Most answers here cover the big three (SQLAlchemy 2.0, SQLModel, Tortoise), so I'll add one more option for the "I like Tortoise's API but want it faster" crowd — full disclosure, I'm the author: [yara-orm](https://github.com/vsdudakov/yara-orm).

It's a Tortoise-style async ORM (models, querysets, relations, migrations) but the engine — pooling, parameter binding, row decoding — is written in Rust (tokio-postgres + rusqlite via PyO3). PostgreSQL and SQLite. The thing I cared about most wasn't raw speed (the ORM is rarely your bottleneck) but correctness: SQLite foreign keys are actually enforced, decimals round-trip exactly, and migrations do auto rename-detection + partial indexes out of the box, which is where Aerich tends to hurt.

Not pitching it over SQLAlchemy for large teams — just a lighter, async-first alternative if Tortoise's ergonomics appeal but its migration story doesn't. Happy to answer questions.
